CMake/Findmlpack.cmake

```diff
@@ -0,0 +1,641 @@
+# - Find mlpack
+# Find the mlpack C++ library
+#
+# author Omar Shrit
+#
+# This module will look for mlpack dependencies and then look for mlpack itself
+#
+# This module can also fetch mlpack dependencies from online repositories.
+#
+# This module sets the following variables:
+#  MLPACK_FOUND - set to true if the library is found
+#  MLPACK_VERSION_MAJOR - major version number
+#  MLPACK_VERSION_MINOR - minor version number
+#  MLPACK_VERSION_PATCH - patch version number
+#  MLPACK_VERSION_STRING - version number as a string (ex: "1.0.4")
+#  MLPACK_INCLUDE_DIRS - list of required include directories
+#  MLPACK_LIBRARIES - list of required libraries
+
+
+## What needs to be done in this PR:
+# 3. If found make it set a list of one that are not found
+# 4. Use the list in the case we need to call the autodownloader
+# 5. Simplyfy the Autodownloader call and make it only one with for loop if
+# necessary, or even better, make all the link recorded already inside the
+# autodownload function and then iterate over the current ones.
+# This will simplfy the autodownloader cmake file
+# STB locally and no need to search for it in the case of autdownload.
+# Be sure to have Autdownload in the same file, maybe even move autodownload
+# into this one.
+# Add another function fetch_mlpack() from source code using autdownloader
+# this should happen inside this file, the user does not need to call
+# autdownload or anyting else
+# A simple comment above this function call should be enough
+# another function can be called fetch_mlpack_and_crosscompile()
+# move minimum version to this file since it is not required anyware else.
+
+# The purpose of this file is to allow the user to use CMake to find mlpack in
+# any usecase. Therefore, some of the parameters are moved from inside
+# CMakeLists to here, since they will be required to be set by the user at
+# somepoint.
+
+##===================================================
+##  MLPACK DEPEDENCIES SETTINGS. 
+##===================================================
+
+# Set minimum library versions required by mlpack.
+#
+# For Armadillo, try to keep the minimum required version less than or equal to
+# what's available on the current Ubuntu LTS or most recent stable RHEL release.
+# See https://github.com/mlpack/mlpack/issues/3033 for some more discussion.
+set(ARMADILLO_VERSION "10.8.2")
+set(ENSMALLEN_VERSION "2.10.0")
+set(CEREAL_VERSION "1.1.2")
+
+# Set required standard to C++17.
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+set(MLPACK_DISABLE_OPENMP OFF)
+
+if (NOT USE_OPENMP)
+  set(MLPACK_DISABLE_OPENMP ON)
+endif()
+
+##===================================================
+##  MLPACK AUTODOWNLOADER DEPEDENCIES FUNCTIONS
+##===================================================
+
+# This function auto-downloads mlpack dependencies.
+# You need to pass the LINK to download from, the name of
+# the dependency, and the name of the compressed package such as
+# armadillo.tar.gz
```

Comment:
By `name of the compressed package` I guess you mean `filename to store the downloaded link to`?

Reply:
Also worth pointing out for clarity that thinks are downloaded to `${CMAKE_BINARY_DIR}/deps/${PACKAGE}` I suppose?